### PR TITLE
koord-scheduler:refine gang restful api for coscheduling

### PR DESCRIPTION
**Pull request by buptcozy**
### V. Checklist

- [ ] I have written necessary docs and comments
- [ ] I have added necessary unit tests and integration tests
- [ ] All checks passed in `make test`

### CODE REVIEW

File: pkg/scheduler/plugins/coscheduling/core/gang_marshal.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+package core
+
+import (
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+type GangMarshal struct {
```
Comment (eahydra) [inline pkg/scheduler/plugins/coscheduling/core/gang_marshal.go:25]: recommend to move the struct to plugin_service.go, and rename to `GangResponse` or `GangSummary`, `GangDetail`

File: pkg/scheduler/plugins/coscheduling/core/gang_marshal.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+	WaitTime                 time.Duration
+	CreateTime               time.Time
+	Mode                     string
+	MinRequiredNumber        int
+	TotalChildrenNum         int
+	GangGroup                []string
+	Children                 sets.String
+	WaitingForBindChildren   sets.String
+	BoundChildren            sets.String
+	OnceResourceSatisfied    bool
```
Comment (eahydra) [inline pkg/scheduler/plugins/coscheduling/core/gang_marshal.go:36]: Consider making these names easier to read. ResourceSatisfied?

File: pkg/scheduler/plugins/coscheduling/core/gang_marshal.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+package core
+
+import (
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+type GangMarshal struct {
+	Name                     string
```
Comment (eahydra) [inline pkg/scheduler/plugins/coscheduling/core/gang_marshal.go:26]: please add tags for these fields.  and use the style `camelCase`

File: pkg/scheduler/plugins/coscheduling/core/gang_marshal.go [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+	BoundChildren            sets.String
+	OnceResourceSatisfied    bool
+	ScheduleCycleValid       bool
+	ScheduleCycle            int
+	ChildrenScheduleRoundMap map[string]int
+	GangFrom                 string
+	HasGangInit              bool
+}
+
+func NewGangMarshal() *GangMarshal {
```
Comment (eahydra) [inline pkg/scheduler/plugins/coscheduling/core/gang_marshal.go:44]: rename to NewGangResponse/NewGangDetail
and the protocol can be changed to `NewGangDetail(gang *core.Gang) *GangDetail`

**Review by eahydra — CHANGES_REQUESTED** [review]

File: pkg/scheduler/plugins/coscheduling/core/gang.go [hunk 5]
```diff
@@ -95,6 +95,44 @@ func NewGang(gangName string) *Gang {
 	}
 }
 
+func (gang *Gang) Marshal() *GangMarshal {
```
Comment (hormes) [inline pkg/scheduler/plugins/coscheduling/core/gang.go:98]: cool, A separate output struct is better,  remove the json tags in the original struct?

File: pkg/scheduler/plugins/coscheduling/core/gang_marshal.go (hunk 2 above)
Comment (buptcozy) [reply pkg/scheduler/plugins/coscheduling/core/gang_marshal.go:36]: OnceResourceSatisfied means has been satisfied once in history, cause when satified then pod is evicted

File: pkg/scheduler/plugins/coscheduling/core/gang_marshal.go (hunk 3 above)
Comment (buptcozy) [reply pkg/scheduler/plugins/coscheduling/core/gang_marshal.go]: fixed

File: pkg/scheduler/plugins/coscheduling/core/gang_marshal.go (hunk 1 above)
Comment (buptcozy) [reply pkg/scheduler/plugins/coscheduling/core/gang_marshal.go]: fixed

File: pkg/scheduler/plugins/coscheduling/core/gang_marshal.go (hunk 4 above)
Comment (buptcozy) [reply pkg/scheduler/plugins/coscheduling/core/gang_marshal.go]: fixed

File: pkg/scheduler/plugins/coscheduling/core/gang.go (hunk 5 above)
Comment (buptcozy) [reply pkg/scheduler/plugins/coscheduling/core/gang.go]: fixed

File: pkg/scheduler/plugins/coscheduling/plugin_service.go [hunk 6]
```diff
@@ -2,25 +2,94 @@ package coscheduling
...
+		WaitingForBindChildren:   sets.NewString(),
+		BoundChildren:            sets.NewString(),
+		ChildrenScheduleRoundMap: make(map[string]int),
+	}
+
+	if gang == nil {
+		return gangSummary
+	}
+
+	gang.Lock()
```
Comment (eahydra) [inline pkg/scheduler/plugins/coscheduling/plugin_service.go:49]: 应该不需要这个Lock吧.  gang已经是deepcopy出来的。

Comment (buptcozy) [reply pkg/scheduler/plugins/coscheduling/plugin_service.go:48]: deep copy pod没必要，已经把原来的deepcopy去掉了

**Review by eahydra — COMMENTED** [review]
/lgtm

**Review by eahydra — APPROVED** [review]
/lgtm

**Review by yihuifeng — COMMENTED** [review]
/lgtm

**Pull request merged**